add onPrev, onNext, and onClose callbacks

azul opened the pull request:
Apps using the viewer might want to react to changes in the viewer.

In particular make use of the callbacks
to updating the url in the files app when navigating between files.
This way the url can be copied and used to navigate to one particular image in a collection.

It also can be used to add browser history entries.
For this first call the callback and then update the title.
This way the callback can create a new history entry
and the title of that entry will then be updated.

`onPrev` and `onNext` callbacks receive the old and the current `fileInfo`
as arguments.
`onClose` callback does not receive any arguments as of now.

Pull Request consists of two commits:
1. introduce the callbacks
2. make use of them when defining the `actionHandler`.

This pull request is a result of the discussion in #488

skjnldsv commented on src/views/Viewer.vue:
```diff
@@ -555,9 +559,27 @@ export default {
+					actionHandler: (name, context) => {
+						const pushToHistory = ({ fileid }) => {
+							const params = OC.Util.History.parseUrlQuery()
+							const dir = params.dir
+							delete params.dir
+							delete params.fileid
+							params.openfile = fileid
+							const query = 'dir=' + OC.encodePath(dir) + '&' + OC.buildQueryString(params)
+							OC.Util.History.pushState(query)
+						}
 						// replace potential leading double slashes
-						OCA.Viewer.open(`${dir}/${name}`.replace(/^\/\//, '/'))
+						const path = `${context.dir}/${name}`.replace(/^\/\//, '/')
+						const oldQuery = location.search.replace(/^\?/, '')
+						const onClose = () => OC.Util.History.pushState(oldQuery)
+						if (!context.fileInfoModel && context.fileList) {
+							context.fileInfoModel = context.fileList.getModelForFile(name)
+						}
+						if (context.fileInfoModel) {
+							pushToHistory({ fileid: context.fileInfoModel.get('id') })
+						}
```
Please put this whole logic into another file or service :)
This is a bit too messy.


Also, don't use `OC.encodePath`, or `OC.buildQueryString`.
We have dedicated libraries for this ;)
https://docs.nextcloud.com/server/latest/developer_manual/app/javascript-apis.html#npm-packages

azul replied:
- [x] Move the logic into `src/services/FilesActionHandler`.
- [x] Replace `OC.encodePath`
- [ ] ~~Replace `OC.buildQueryString`~~ -  did not find a replacement. Is also not deprecated as of now.

skjnldsv commented on src/views/Viewer.vue:
```diff
@@ -674,6 +710,18 @@ export default {
+			this.Viewer.onPrev(info)
+		},
+
+		onNext(info) {
```
You're missing one arg

azul replied:
thanks!

azul commented on src/views/Viewer.vue:
```diff
@@ -222,7 +223,7 @@ export default {
 				console.info('Opening viewer for file ', path)
 				this.openFile(path)
-			} else {
+			} else if (this.initiated) {
```
oh... one more thing. I had to add this condition because the close function was getting twice:
* clicking the :x: triggers the close function
* the close function sets file to `''` which triggers this observer fuction.

Since `close` also sets `this.initiated` to false this condition works to prevent calling close again. However it seems like a workaround for a more fundamental problem. 

Should i try and solve the underlying problem as part of this pr?
What would be the best way to approach it?

skjnldsv replied:
Yes, if you can  narrow why is it called twice, that would be better I think :)
Thanks!

azul replied:
I figured out what happens:
* clicking the :x: triggers the close function
* close function calls `OCA.Viewer.close()` and triggers the callback
* `OCA.Viewer.close()` sets `file` to `''`
* the `file` watcher triggers the close function again
* close function calls  `OCA.Viewer.close()` and triggers the callback again
* `OCA.Viewer.close()` sets `file` to `''`
* this time it has no effect because it already was empty.

azul replied:
In order to fix this i separate `cleanup` from the `close` function
  
* `close()` delegates to `OCA.Viewer.close()`.
    It is just a shortcut to be used in the template.
    
* `cleanup()` does the actual work.
    `OCA.Viewer.close()` will set `file` to `''`.
    `file` is being watched and the watch function triggers the cleanup.

azul commented:
@skjnldsv If you could take another look... that would be great :smiley:

skjnldsv commented:
> @skjnldsv If you could take another look... that would be great

Sorry, I'm overwhelmed since a few months! :(
I definitely did not forget about this pr :)
Thanks for the heads up!!

azul commented:
> Sorry, I'm overwhelmed since a few months! :(

That's what i was assuming. I know how it feels. Take your time. :tea: :smiling_face_with_three_hearts:

skjnldsv commented on src/views/Viewer.vue:
```diff
@@ -447,6 +448,10 @@ export default {
 		},
 
+		updateTitle(fileName) {
+			document.title = `${fileName} - ${OC.theme.title}`
```
```suggestion
			document.title = `${fileName} - ${OCA.Theming.name}`
```

skjnldsv commented on src/services/Viewer.js:
```diff
@@ -90,7 +93,7 @@ export default class Viewer {
 	 * @param {string} path the path to open
 	 * @param {Object[]} [list] the list of files as objects (fileinfo) format
```
Could you also update the jsdoc, it is outdated and doesn't reflect the Object params requirements (I forgot to update it before, but since we're at it :see_no_evil:)

skjnldsv commented on src/services/Viewer.js:
```diff
@@ -90,7 +93,7 @@ export default class Viewer {
 	 * @param {Object[]} [list] the list of files as objects (fileinfo) format
 	 */
-	open({ path, list = [], loadMore = () => ([]) } = {}) {
+	open({ path, list = [], loadMore = () => ([]), onPrev = () => {}, onNext = () => {}, onClose = () => {} }) {
```
```suggestion
	open({ path, list = [], loadMore = () => ([]), onPrev = () => {}, onNext = () => {}, onClose = () => {} } = {}) {
```

skjnldsv commented:
Please also rebase! :)
A few comments and reviews, let's try to get this in soon so you don't feel frustrated by the wait! :wink: 
Thanks again for your patience!! :hugs:

skjnldsv commented:
Let's rebase, squash and get this in? :)

azul commented:
@skjnldsv I'm in a call right now. This is the next thing I'll work on.

azul commented:
@skjnldsv I rebased and updated the jsdoc.
I'm not so familiar with jsdoc syntax. So please take a look if it's correct :eyeglasses: 
Shall I squash everything into a single commit?

skjnldsv commented:
> Shall I squash everything into a single commit?

Yes please :)

skjnldsv commented:
Btw @azul , this also fixes https://github.com/nextcloud/viewer/pull/497, right?

azul commented:
:heavy_check_mark: squashed
:heavy_check_mark: rebased again

> Btw @azul , this also fixes #497, right?

This **is** #497 - I tried to find the issue you meant to point at but did not see it.

@skjnldsv - ready to merge?

skjnldsv commented:
> This **is** #497 - I tried to find the issue you meant to point at but did not see it.

sorry, #486 :grin:

skjnldsv commented:
/compile amend /

azul commented:
@skjnldsv It's a first step towards solving #486. Now we add browser history entries for the files we navigate to. But the files app still needs a handler for dealing with clicks to the browser back button that make the viewer display the file. I'll work on that as one of the next things.

skjnldsv commented on src/views/Viewer.vue:
```diff
@@ -674,6 +695,18 @@ export default {
+		},
+
+		onNext(info, oldFileInfo) {
+			this.Viewer.onNext(oldFileInfo)
```
Hey @azul, I just rezlized we did a mistake, the onPrev and onNext have different parametters.
It should most likely be consistent:
```js
this.Viewer.onNext(oldFileInfo)
```
